### PR TITLE
MYRIAD-154 Restarting RM starts a "default" medium-profile NM on each…

### PR DESCRIPTION
**Pull request by sdaingade**
… restart.

We check if there is atleast one NM in pending, staging or active state. if yes,
do not launch additional NMs.

Tested, NM was launched the first time as there were no NMs in pending, staging or active state.
Restarted RM multiple times. No new NM was launched.

### CODE REVIEW

File: myriad-scheduler/src/main/java/com/ebay/myriad/Main.java [hunk 1]
```diff
@@ -211,7 +215,32 @@ private void startNMInstances(Injector injector) {
       Map<String, Integer> nmInstances = injector.getInstance(MyriadConfiguration.class).getNmInstances();
       MyriadOperations myriadOperations = injector.getInstance(MyriadOperations.class);
       ServiceProfileManager profileManager = injector.getInstance(ServiceProfileManager.class);
+      SchedulerState schedulerState = injector.getInstance(SchedulerState.class);
+
+      Set<NodeTask> launchedNMTasks = new HashSet<>();
+      launchedNMTasks.addAll(
+        schedulerState.getPendingTasksByType(NodeManagerConfiguration.NM_TASK_PREFIX));
+      if (!launchedNMTasks.isEmpty()) {
```
Comment (yufeldman) [inline myriad-scheduler/src/main/java/com/ebay/myriad/Main.java]: Let's check if it is < nmInstances and if >= return

File: myriad-scheduler/src/main/java/com/ebay/myriad/Main.java [hunk 2]
```diff
@@ -211,7 +215,32 @@ private void startNMInstances(Injector injector) {
...
+        LOGGER.info("{} NM(s) in pending state. Not launching additional NMs", launchedNMTasks.size());
+        return;
+      }
+
+      launchedNMTasks.addAll(
+        schedulerState.getStagingTasksByType(NodeManagerConfiguration.NM_TASK_PREFIX));
+      if (!launchedNMTasks.isEmpty()) {
```
Comment (yufeldman) [inline myriad-scheduler/src/main/java/com/ebay/myriad/Main.java]: Let's check if it is < nmInstances and if >= return

File: myriad-scheduler/src/main/java/com/ebay/myriad/Main.java [hunk 3]
```diff
@@ -211,7 +215,32 @@ private void startNMInstances(Injector injector) {
...
+        LOGGER.info("{} NM(s) in staging state. Not launching additional NMs", launchedNMTasks.size());
+        return;
+      }
+
+      launchedNMTasks.addAll(
+        schedulerState.getActiveTasksByType(NodeManagerConfiguration.NM_TASK_PREFIX));
+      if (!launchedNMTasks.isEmpty()) {
```
Comment (yufeldman) [inline myriad-scheduler/src/main/java/com/ebay/myriad/Main.java]: Same as above

File: myriad-scheduler/src/main/java/com/ebay/myriad/Main.java [hunk 4]
```diff
@@ -211,7 +215,32 @@ private void startNMInstances(Injector injector) {
...
+        LOGGER.info("{} NM(s) in active state. Not launching additional NMs", launchedNMTasks.size());
+        return;
+      }
+
       for (Map.Entry<String, Integer> entry : nmInstances.entrySet()) {
```
Comment (yufeldman) [inline myriad-scheduler/src/main/java/com/ebay/myriad/Main.java]: Let's launch the difference

File: myriad-scheduler/src/main/java/com/ebay/myriad/Main.java (hunk 1 above)
Comment (sdaingade) [inline myriad-scheduler/src/main/java/com/ebay/myriad/Main.java]: Currently I am checking if there is at least one NM in pending, staging or active state on restart (irrespective of profile type). If this is the first time the cluster is being launched, there will be no NM running. In this case we look at the myriad-default-config.xml to look at number of NMs to be launched per profile.

**Comment by smarella** [comment]
Ship It!

**Pull request closed without merging**